common/flatpak-run.c

```diff
@@ -3322,10 +3323,88 @@ flatpak_run_app (const char     *app_ref,
 
   if (app_deploy != NULL)
     {
+      g_autofree const char **previous_ids = NULL;
+      gsize len = 0;
+      int i;
+
+      real_app_id_dir = flatpak_get_data_dir (app_ref_parts[1]);
+
+      previous_app_id_dirs = g_ptr_array_new_with_free_func (g_object_unref);
       app_files = flatpak_deploy_get_files (app_deploy);
 
-      real_app_id_dir = flatpak_ensure_data_dir (app_ref_parts[1], cancellable, error);
-      if (real_app_id_dir == NULL)
+      previous_ids = flatpak_deploy_data_get_previous_ids (app_deploy_data, &len);
+      for (i = len - 1; i >= 0; i--)
+        {
+          g_autoptr(GFile) previous_app_id_dir = NULL;
+          g_autoptr(GFileInfo) previous_app_id_dir_info = NULL;
+          g_autoptr(GError) local_error = NULL;
+          gboolean create_symlink;
+
+          previous_app_id_dir = flatpak_get_data_dir (previous_ids[i]);
+          previous_app_id_dir_info = g_file_query_info (previous_app_id_dir,
+                                                        G_FILE_ATTRIBUTE_STANDARD_IS_SYMLINK ","
+                                                        G_FILE_ATTRIBUTE_STANDARD_SYMLINK_TARGET,
+                                                        G_FILE_QUERY_INFO_NOFOLLOW_SYMLINKS,
+                                                        cancellable,
+                                                        &local_error);
+          if (previous_app_id_dir_info == NULL && !g_error_matches (local_error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND))
+            {
+              g_propagate_error (error, g_steal_pointer (&local_error));
+              return FALSE;
+            }
+
+          g_clear_error (&local_error);
+
+          /* If the previous data dir exists, but the current one doesn't, rename the old
+             one to the new one. (We need to be careful to not move some old symlink over
+             onto the current directory.) */
+          if (previous_app_id_dir_info != NULL && !g_file_info_get_is_symlink (previous_app_id_dir_info) &&
+              !g_file_query_exists (real_app_id_dir, cancellable))
+            {
+              if (!flatpak_file_rename (previous_app_id_dir, real_app_id_dir, cancellable,
+                                        &local_error))
+                {
+                  g_propagate_error (error, g_steal_pointer (&local_error));
+                  return FALSE;
+                }
+
+              /* Treat it as nonexistent, since it now isn't anyway. */
+              g_clear_object (&previous_app_id_dir_info);
+            }
+
+          /* Each previous data dir should symlink to the new one. If a symlink is already
+             present but not pointing to the new location, it needs to be removed and updated.
+
+             However, if the user somehow ended up with two data directories, and only one is
+             moved over, then the second should be left alone, since there's not much we can
+             really do at that point. */
```

Comment:
I think this stuff is unnecessary risky. We should not need to modify any existing symlinks, if we have a setup like:
a -> b
b: dir

and then b is rebased to c, then we will be fine with this outcome:
a -> b
b -> c
c: dir

But, if we start tweaking existing symlinks we might run into problems if the user did something weird.
Basically, we should *only* need to create the back-compat symlink when actually renaming some existing file to the new directory.

For example, if we have app c, with previous names a and b, but we only ran it with name a, then we should rename a to c and add a a->c symlink. Nothing ever ran with the b name, so it doesn't need a symlink. Similarly if we have a b directory we need only rename b to c and add a b symlink. If we previously ran with the a name, then there will be a a->b link, and if we didn't we don't need one.